File: Proj1/src/main.cpp
#include "matrix.hpp"
#include "papi.hpp"

#include <iostream>
#include <cstdlib>
#include <sstream>
#include <math.h>
#include <ctime>

#define M 1000000
#define MX 1024

using namespace std;

size_t size_A;
size_t size_B;

static bool right_args(int argc, char const *argv[]);
static bool right_sizes(char const *argv[]);


int main(int argc, char const *argv[]) {
    Papi papi = Papi();

    papi.Init();
    papi.InstallEvents();

    if(!right_args(argc, argv))
        exit(1);

    if(!right_sizes(argv))
        exit(1);

    float tot_FLOPS = 2*MX*MX*size_A;

    Matrix *matrix_A = new Matrix(size_A, size_A);
    matrix_A->fill(matrix_A);

    Matrix *matrix_B = new Matrix(size_B, size_B);
    matrix_B->fill(matrix_B);

    cout << "Basic Multiplication" << endl;

    clock_t begin = clock();

    papi.Start();

    // Gets the start time in clock cycles
    long long basic_start_cycles = papi.GetRealCycles();

    // Gets the start time in microseconds cycles
    long long basic_start_usec = papi.GetRealuSec();

    // Matrix *basic_result = matrix_A->basic_multiply(matrix_A, matrix_B);
    matrix_A->basic_multiply(matrix_A, matrix_B);

    // Gets the end time in clock cycles
    long long basic_end_cycles = papi.GetRealCycles();

    // Gets the start time in microseconds cycles
    long long basic_end_usec = papi.GetRealuSec();

    papi.StopAndReset();

    clock_t end = clock();

    cout << "MFlop/s = " << (tot_FLOPS/M)/(end-begin) << endl
         << "Time (clock cycles) = "
         << (basic_end_cycles - basic_start_cycles) << endl
         << "Time (seconds) = "
         << (basic_end_usec - basic_start_usec)*pow(10,-6) << endl
         << endl;

    cout << "Line Multiplication" << endl;

    clock_t begin2 = clock();

    papi.Start();

    // Gets the start time in clock cycles
    long long line_start_cycles = papi.GetRealCycles();

    // Gets the start time in microseconds cycles
    long long line_start_usec = papi.GetRealuSec();

    //Matrix *line_result = matrix_A->line_multiply(matrix_A, matrix_B);
    matrix_A->line_multiply(matrix_A, matrix_B);

    // Gets the end time in clock cycles
    long long line_end_cycles = papi.GetRealCycles();

    // Gets the start time in microseconds cycles
    long long line_end_usec = papi.GetRealuSec();

    papi.StopAndReset();

    clock_t end2 = clock();

    cout << "MFlop/s = " << (tot_FLOPS/M)/(end2-begin2) << endl
         << "Time (clock cycles) = "
         << (line_end_cycles - line_start_cycles) << endl
         << "Time (seconds) = "
         << (line_end_usec - line_start_usec)*pow(10,-6) << endl
         << endl;

    // Test Matrix Values
    /*cout << "Matrix A" << endl;
    matrix_A->print(matrix_A);
    cout << "Matrix B" << endl;
    matrix_A->print(matrix_B);
    cout << "Matrix AxB [BASIC]" << endl;
    matrix_A->print(basic_result);
    cout << "Matrix AxB [LINE]" << endl;
    matrix_A->print(line_result);*/

    return 0;
}

static bool right_args(int argc, char const *argv[]) {
    if(argc != 3) {
        cerr << "[Wrong_Args] usage: " << argv[0] << " "
             << "<size_matrix_A> <size_matrix_B>" << endl
             << "\t Where:" << endl
             << "\t\t <size_matrix_A> - matrix size_matrix_A x size_matrix_A"
             << endl
             << "\t\t <size_matrix_B> - matrix size_matrix_B x size_matrix_B"
             << endl;
        return false;
    }

    return true;
}

static bool right_sizes(char const *argv[]) {
    istringstream s_A(argv[1]);
    istringstream s_B(argv[2]);

    if(!(s_A >> size_A) || !(s_B >> size_B)) {
        cerr
            << "[Wrong_Args] <size_matrix_A> and <size_matrix_B> must be integers"
            << endl;
        return false;
    }

    return true;
}

File: Proj1/py/src/main.py
import sys
import random
import time
from matrix import Matrix, basic_multiply, line_multiply
from pypapi import papi_high, events as papi_events

size_A = 0
size_B = 0
algorithm = 0
M = 1000000
MX = 1024

def main():
    if not(right_args(len(sys.argv), sys.argv)):
        return 1

    size_A = int(sys.argv[2])
    size_B = int(sys.argv[3])
    tot_FLOPS = 2*MX*MX*size_A;

    if valid_algorithm(sys.argv[1]):
        algorithm = int(sys.argv[1])
    else:
        return 2

    matrix_A = Matrix(size_A, size_A)
    fill(matrix_A)

    matrix_B = Matrix(size_B, size_B)
    fill(matrix_B)

    # Test Matrix Values
    #print_matrix(matrix_A)
    #print_matrix(matrix_B)

    # Starts some counters
    papi_high.start_counters([papi_events.PAPI_L1_TCM, papi_events.PAPI_L2_TCM, papi_events.PAPI_TOT_INS])

    if algorithm == 1:
        print 'Basic Matrix Multiplication'
        start_time = time.time()
        result = basic_multiply(matrix_A, matrix_B)
        end_time = time.time()
        #print_matrix(result)

    if algorithm == 2:
        print 'Line Matrix Multiplication'
        start_time = time.time()
        result = line_multiply(matrix_A, matrix_B)
        end_time = time.time()
        #print_matrix(result)

    # Reads values from counters and reset them
    results = papi_high.read_counters()  # -> [int, int]

    # Print results
    print_times(results, tot_FLOPS, start_time, end_time)

    # Stop counters
    papi_high.stop_counters()  # -> []


def right_args(num_args, argv):
    if num_args != 4:
        print '[Wrong_Args] usage: ', argv[0], ' ', '<algorithm> <size_matrix_A> <size_matrix_B>\n'
        print '\t Where:\n'
        print '\t\t <algorithm> - type of algorithm to apply\n'
        print '\t\t <size_matrix_A> - matrix size_matrix_A x size_matrix_A\n'
        print '\t\t <size_matrix_B> - matrix size_matrix_B x size_matrix_B\n'
        return False

    return True

def valid_algorithm(alg):
    if (int(alg) == 1) or (int(alg) == 2):
        return True
    else:
        print '[Wrong_Algorithm] You must choose one of the following options:'
        print '\t1 - Basic Matrix Multiplication'
        print '\t2 - Line Matrix Multiplication'
        return False

def fill(matrix):
    for i in range(matrix.num_rows*matrix.num_cols):
        matrix.matrix_values[i] = random.randint(1,100)

def print_matrix(matrix):
    print '[',
    for i in range(matrix.num_rows*matrix.num_cols):
        print matrix.matrix_values[i] ,
    print ']'

def print_times(results, tot_FLOPS, start_time, end_time):
    print 'PAPI_L1_TCM = ', results[0]
    print 'PAPI_L2_TCM = ', results[1]
    print 'PAPI_TOT_INS = ', results[2]
    print 'MFLOP/s = ', (tot_FLOPS/M)/(end_time-start_time)
    print 'Time (seconds) = ', (end_time-start_time)

if __name__ == "__main__":
    main()
